### PR TITLE
Add support for arbitrary args in keybindings

### LINKED ISSUE #1142: Feature Request: Support arbitrary arguments for keybindings
_From discussion with @carlos-zamora and @DHowett-MSFT, and possibly in #968_

Keybindings need to support arbitrary args, so that the handlers can parse their own arg values.

```
{
    keys: ["ctrl+shift+c"],
    action: copy,
    args:{
        type: "mungeData"
    }
},
{
    keys: ["ctrl+shift+pgup"],
    action: scroll,
    args:{
        amount: -1
    }
}
```

This is a feature that needs to absolutely be spec'd before implemented.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Enables the user to provide arbitrary argument values to shortcut actions through a new `args` member of keybindings. For some keybindings, like `NewTabWithProfile<N>`, we previously needed 9 different `ShortcutAction`s, one for each value of `Index`. If a user wanted to have a `NewTabWithProfile11` keybinding, that was simply impossible. Now that the args are in their own separate json object, each binding can accept any number of arbitrary argument values.

So instead of:
```json
        { "command": "newTab", "keys": ["ctrl+shift+t"] },
        { "command": "newTabProfile0", "keys": ["ctrl+shift+1"] },
        { "command": "newTabProfile1", "keys": ["ctrl+shift+2"] },
        { "command": "newTabProfile2", "keys": ["ctrl+shift+3"] },
        { "command": "newTabProfile3", "keys": ["ctrl+shift+4"] },
```

We can now use:

```json
        { "command": "newTab", "keys": ["ctrl+shift+t"] },
        { "command": { "action": "newTab", "index": 0 }, "keys": ["ctrl+shift+1"] },
        { "command": { "action": "newTab", "index": 1 }, "keys": ["ctrl+shift+2"] },
        { "command": { "action": "newTab", "index": 2 }, "keys": ["ctrl+shift+3"] },
```

Initially, this does seem more verbose. However, for cases where there are multiple args, or there's a large range of values for the args, this will quickly become a more powerful system of expressing keybindings.

The "legacy" keybindings are _left in_ in this PR. They have helper methods to generate appropriate `IActionArgs` values. Prior to releasing 1.0, I think we should remove them, if only to remove some code bloat.

## References

See [the spec](https://github.com/microsoft/terminal/blob/master/doc/specs/%231142%20-%20Keybinding%20Arguments.md) for more details.

This is part two of the implementation, part one was #2446

## PR Checklist
* [x] Closes #1142
* [x] I work here
* [x] Tests added/passed
* [x] Schema updated

## Validation Steps Performed

* Ran Tests
* Removed the legacy keybindings from the `defaults.json`, everything still works
* Tried leaving the legacy keybingings in my `profiles.json`, everything still works.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
I hate to suggest this, but.. is args _part of_ command? Should it be? if it is, we could double-parse the command object for the name _and_ the args:


```
"command": {
    "name": "newTab",
    "index": 4
}
```

command.name = look up actionargs type, actargs.FromJson(commandObject)?

**Comment by DHowett-MSFT** [comment]
That might make the model a little easier to understand when we expand it to "command palette", binding menu items to "commands", etc. The command has all the data necessary to _act_

**Comment by zadjii-msft** [comment]
@DHowett-MSFT another idea: we just put them right in the root directly. 

```
            { "command": "switchToTab", "keys": [ "alt+shift+1" ], "index": 2 },
            { "command": "splitPane", "keys": [ "alt+shift+|" ], "type": "vertical", "profileGuid": "{etc}" },

```

**Comment by DHowett-MSFT** [comment]
Interesting! Will we ever have a command that has a `keys`?

Like:

```
{ "command": "sendKeys", "keys": ["alt+left", "ctrl+q"], "keys": ["shift-1"] }
```

Obviously, activated with <kbd>Shift+1</kbd> and when you activate it it prints <kbd>Alt+Left</kbd>,<kbd>Ctrl+Q</kbd> into the input buffer. 😁

**Comment by zadjii-msft** [comment]
Frick. That's a good point. Lets loop back on the previous proposal then:

```json
    "keybindings":
    [
        { "command": "scrollUpPage", "keys": ["ctrl+shift+pgup"] },
        { "command": { "action": "switchToTab", "index": 0 }, "keys": ["ctrl+alt+1"] },
        { "command": { "action": "switchToTab", "index": 1 }, "keys": ["ctrl+alt+2"] },
        { "command": { "action": "sendInput", "keys": ["alt+left", "ctrl+q"] }, "keys": ["shift+1"] }
    ]
```

If "command" is a string, we'll assume it doesn't have args, and we'll try and parse it as the action.

Otherwise, we'll try and use the "action" property to get the ShortcutAction.

I'm proposing "action" here, because for #2046, commands will need another separate name property that we can use for displaying:


```json
    "commands":
    [
        { 
            "name": "Scroll up a page", 
            "icon":null,
            "command": "scrollUpPage" 
        },
        { 
            "name": "Switch to tab 0", 
            "icon":null,
            "command": { "action": "switchToTab", "index": 0 } 
        },
        { 
            "name": "Switch to the second tab", 
            "icon":null,
            "command": { "action": "switchToTab", "index": 1 } 
        },
        { 
            "name": "Do a send input", 
            "icon":null,
            "command": { "action": "sendInput", "keys": ["alt+left", "ctrl+q"] }
         }
    ]
```

Oh no. Now I hate myself. Why do commands and keybindings need to be in separate arrays at all? (I'm putting this on in an expando because it's getting long)

<details>

```jsonc
    "commands":
    [
        { 
            "name": "Scroll up a page", 
            "icon":null,
            "command": "scrollUpPage",
            "keys": ["ctrl+shift+pgup"]
        },
        { 
            "name": "Switch to tab 0", 
            "icon":null,
            "command": { "action": "switchToTab", "index": 0 },
            "keys": ["ctrl+alt+1"]
        },
        { 
            "name": "Switch to the second tab", 
            "icon":null,
            "command": { "action": "switchToTab", "index": 1 },
            "keys": ["ctrl+alt+2"]
        },
        { 
            "name": "Do a send input", 
            "icon":null,
            "command": { "action": "sendInput", "keys": ["alt+left", "ctrl+q"] },
            "keys": ["shift+1"]
         },
        { 
            // I'm just a keybinding, not a command
            "command": { "action": "sendInput", "keys": ["alt+right", "ctrl+w"] },
            "keys": ["shift+2"]
        },
        { 
            "name": "I'm just a command, unbound", 
            "icon":null,
            "command": { "action": "sendInput", "keys": ["ctrl+left", "ctrl+1"] }
        },
        { 
            "name": { "key": "SomeLocalizedXAMLStringResource"}, 
            "icon":null,
            "command": { "action": "switchToTab", "index": 15 },
            "keys": ["ctrl+alt+5"]
        }
    ]
```

</details>

We'd just have all commands and bindings in a singular array.

**Comment by DHowett-MSFT** [comment]
I would _love_ to have this followed up with a PR that adds `profileGuid` to `newTab` :grin:

**Comment by zadjii-msft** [comment]
Looping back on this - Dustin and I discussed, and the combined commands/keybindings situation is a madhouse, which is part of the reason we're leaving it for 2.0.

However, there was goodness we agreed on, which is the following:
<hr>


```json
    "keybindings":
    [
        { "command": "scrollUpPage", "keys": ["ctrl+shift+pgup"] },
        { "command": { "action": "switchToTab", "index": 0 }, "keys": ["ctrl+alt+1"] },
        { "command": { "action": "switchToTab", "index": 1 }, "keys": ["ctrl+alt+2"] },
        { "command": { "action": "sendInput", "keys": ["alt+left", "ctrl+q"] }, "keys": ["shift+1"] }
    ]
```

If "command" is a string, we'll assume it doesn't have args, and we'll try and parse it as the action.

Otherwise, we'll try and use the "action" property to get the ShortcutAction.

<hr>

**Comment by miniksa** [comment]
> The "legacy" keybindings are _left in_ in this PR. They have helper methods to generate appropriate `IActionArgs` values. Prior to releasing 1.0, I think we should remove them, if only to remove some code bloat.

TODO without work item ID attached.

File: src/cascadia/TerminalApp/ActionArgs.h [hunk 1]
```diff
@@ -36,30 +37,134 @@ namespace winrt::TerminalApp::implementation
     {
         CopyTextArgs() = default;
         GETSET_PROPERTY(bool, TrimWhitespace, false);
+
+        static constexpr std::string_view TrimWhitespaceKey{ "trimWhitespace" };
+
+    public:
+        static winrt::TerminalApp::IActionArgs FromJson(const Json::Value& json)
+        {
+            // IMPORTANT: Not using make_self here _will_ break you in the future!
```
Comment (miniksa) [inline src/cascadia/TerminalApp/ActionArgs.h]: would you say it's.... load bearing? :P

File: src/cascadia/TerminalApp/ActionArgs.h [hunk 2]
```diff
@@ -36,30 +37,134 @@ namespace winrt::TerminalApp::implementation
...
+        else if (directionString == UpString)
+        {
+            return TerminalApp::Direction::Up;
+        }
+        else if (directionString == DownString)
+        {
+            return TerminalApp::Direction::Down;
+        }
+        // default behavior for invalid data
+        return TerminalApp::Direction::Left;
```
Comment (miniksa) [inline src/cascadia/TerminalApp/ActionArgs.h]: Uh, in what context is this usually used and is Left an appropriate default?

(That is... is it either sufficiently good for most circumstances such that a null is OK... or alternatively is it sufficiently bad that someone will instantly recognize they made a mistake?)

File: src/cascadia/TerminalApp/defaults.json [hunk 3]
```diff
@@ -1,5 +1,7 @@
 // THIS IS AN AUTO-GENERATED FILE! Changes to this file will be ignored.
 {
+    "$schema": "file:///c:/Users/migrie/dev/private/OpenConsole/doc/cascadia/profiles.schema.json",
```
Comment (miniksa) [inline src/cascadia/TerminalApp/defaults.json]: NOPE.

**Comment by zadjii-msft** [comment]
@zadjii-msft Make sure to update the spec before you submit this PR

File: src/cascadia/TerminalApp/ActionArgs.h [hunk 4]
```diff
@@ -36,30 +37,134 @@ namespace winrt::TerminalApp::implementation
...
+            if (auto tabIndex{ json[JsonKey(TabIndexKey)] })
+            {
+                args->_TabIndex = tabIndex.asInt();
+            }
+            return *args;
+        }
+    };
+
+    // Possible Direction values
+    static constexpr std::string_view LeftString{ "left" };
```
Comment (miniksa) [inline src/cascadia/TerminalApp/ActionArgs.h]: Shouldn't constant string patterns like this for parsing go wherever the rest of the parsing is for things like dimensions?

File: src/cascadia/TerminalApp/AppKeyBindingsSerialization.cpp [hunk 5]
```diff
@@ -227,18 +395,49 @@ void winrt::TerminalApp::implementation::AppKeyBindings::LayerJson(const Json::V
...
+            // { "command": { "action": "switchToTab", "index": 0 }, "keys": ["ctrl+alt+1"] },
+
+            // 1. In the first case, the "command" is a string, that's the
+            //    action name. There are no provided args, so we'll pass
+            //    Json::Value::null to the parse function.
+            // 2. In the second case, the "command" is an object. We'll use the
+            //    "action" in that object as the action name. We'll then pass
+            //    the "command" object to the arg parser, for furhter parsing.
+
+            // auto argsVal = value[JsonKey(ArgsKey)];
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppKeyBindingsSerialization.cpp]: old code?

File: src/cascadia/TerminalApp/AppKeyBindings.h [hunk 6]
```diff
@@ -79,8 +80,9 @@ namespace winrt::TerminalApp::implementation
         // clang-format on
 
     private:
-        std::unordered_map<winrt::Microsoft::Terminal::Settings::KeyChord, TerminalApp::ShortcutAction, KeyChordHash, KeyChordEquality> _keyShortcuts;
-        bool _DoAction(ShortcutAction action);
+        // std::unordered_map<winrt::Microsoft::Terminal::Settings::KeyChord, TerminalApp::ShortcutAction, KeyChordHash, KeyChordEquality> _keyShortcuts;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppKeyBindings.h]: old code?

File: src/cascadia/TerminalApp/ActionAndArgs.cpp [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#include "pch.h"
+#include "ActionAndArgs.h"
+#include "ActionAndArgs.g.cpp"
+
+namespace winrt::TerminalApp::implementation
+{
+}
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ActionAndArgs.cpp]: Do we even need this file?

**Pull request merged**